### DISCUSSION TITLE
Question on private scalar derivation (ed25519)

### DISCUSSION THREAD

**Discussion by matthiasgeihs** [paulmillr/noble-curves#109] [Q&A] [answered]
https://github.com/paulmillr/noble-curves/blob/0a663391bd3446f62f473e9cfd9fb68b5ad5523e/src/abstract/edwards.ts#L424-L427

As far as I understand, the reason for bit clamping ed25519 keys is to mitigate small subgroup attacks. This is achieved by ensuring that the secret scalar is a multiple of 8 (by clamping the 3 lowest order bits).

In line 427, however, the clamped secret scalar is reduced modulo the prime group order (and thereby the bit representation is changed). Doesn't this defeat the purpose of the clamping? Shouldn't the final scalar have the 3 lowest order bits zeroed out? Or am I getting something wrong here?

**Comment by matthiasgeihs** [comment, marked as answer]
I'll try to answer my own question here.

In the case of `ed25519` signatures, this modulo operation does not have an effect, since we know that we are operating on points of the prime order subgroup.

In the case of `x25519` this would have an effect, but here we are not applying the modulo operation.
https://github.com/paulmillr/noble-curves/blob/0a663391bd3446f62f473e9cfd9fb68b5ad5523e/src/abstract/montgomery.ts#L157-L163

**Comment by paulmillr** [comment]
yeah we just follow the spec even if it doesn’t matter much
